### PR TITLE
coder: fix build and update to v2.0.2

### PR DESCRIPTION
**Pull request by chrisportela**
## Description of changes
Updates coder to v2.0.2 and fixes package on linux & darwin

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [x] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by chrisportela** [comment]
There is actually an issue with this. While it does build again, this build is incomplete without the slim binaries being in the right place (`./site/bin`) so the coder server `embed` build can serve up those binaries when provisioning new instances.

**Comment by gmemstr** [comment]
@chrisportela Might make sense to bump this to https://github.com/coder/coder/releases/tag/v2.2.0. Taking a look into the issues - locally, I'm getting an esbuild related error, so have to dig further.

File: pkgs/development/tools/coder/default.nix [hunk 1]
```diff
@@ -42,29 +72,43 @@ buildGoModule rec {
   subPackages = [ "cmd/..." ];
 
   preBuild = ''
-    export HOME=$TEMPDIR
+    export HOME=$(mktemp -d)
+
```
Comment (gmemstr) [inline pkgs/development/tools/coder/default.nix:76]: From testing locally, this allows us to pick up the ESBuild binary

```suggestion
    export HOME=$(mktemp -d)
    export ESBUILD_BINARY_PATH=${ESBUILD_BINARY_PATH}
```

**Comment by gmemstr** [comment]
Regardless of that issue, even if you get it running environments won't actually come up because the webserver doesn't have the binaries to serve up to the containers :/

**Comment by chrisportela** [comment]
> > Regardless of that issue, even if you get it running environments won't actually come up because the webserver doesn't have the binaries to serve up to the containers :/
> 
> Does that affect use as a Coder client, as a Coder server or both? (The same package is used for both use-cases as far as I can tell.)

Just as a server (which is what I cared about most).

I solved my immediate issues by using downloading the binaries from Coder directly: 

```nix
{ lib
, fetchurl
, installShellFiles
, makeWrapper
, terraform
, stdenvNoCC
, unzip
}:
let
  inherit (stdenvNoCC.hostPlatform) system;
in

stdenvNoCC.mkDerivation rec {
  pname = "coder";
  version = "2.1.5";

  src = fetchurl {
    sha256 = {
      x86_64-linux = "sha256-yIhyVxsYfMfsNvMu6wohs/57FXYCnPSibR8l3eu1p3M=";
      x86_64-darwin = lib.fakeHash;
      aarch64-linux = lib.fakeHash;
      aarch64-darwin = "sha256-yDIns1iZ2wEpXBQLoo/krchv+ETL3rpmFbDvAu1VkXA=";
    }.${system};

    url =
      let
        systemName = {
          x86_64-linux = "linux_amd64";
          aarch64-linux = "linux_arm64";
          x86_64-darwin = "darwin_amd64";
          aarch64-darwin = "darwin_arm64";
        }.${system};

        ext = {
          x86_64-linux = "tar.gz";
          aarch64-linux = "tar.gz";
          x86_64-darwin = "zip";
          aarch64-darwin = "zip";
        }.${system};
      in
      "https://github.com/coder/coder/releases/download/v${version}/coder_${version}_${systemName}.${ext}";
  };

  nativeBuildInputs = [
    installShellFiles
    makeWrapper
    unzip
  ];

  unpackPhase = ''
    printf 'Decompressing %s\n' "$src"
    case $src in
        *.tar.gz) tar -xz -f "$src" ;;
        *.zip)    unzip      "$src" ;;
    esac
  '';

  installPhase = ''
    mkdir -p $out/bin
    cp coder $out/bin
  '';

  postInstall = ''
    installShellCompletion --cmd coder \
      --bash <($out/bin/coder completion bash) \
      --fish <($out/bin/coder completion fish) \
      --zsh <($out/bin/coder completion zsh)

    wrapProgram $out/bin/coder --prefix PATH : ${lib.makeBinPath [ terraform ]}
  '';

  # integration tests require network access
  doCheck = false;

  meta = {
    description = "Provision software development environments via Terraform on Linux, macOS, Windows, X86, ARM, and of course, Kubernetes";
    homepage = "https://coder.com";
    license = lib.licenses.agpl3;
    maintainers = [ lib.maintainers.ghuntley lib.maintainers.urandom ];
    broken = false;
  };
}
```

But as a client this PR + updating the version should work totally fine.

**Comment by das-g** [comment]
It'd be nice anyway to be able to use the coder client without pulling in all dependencies of the coder server, especially now that terraform has switched to an unfree license.

**Comment by carloaromando** [comment]
@chrisportela 
I tried to build this `coder` version on my `x86_64-darwin` machine but it fails with this log:
```
> coder-v2@ build /private/tmp/nix-build-coder-2.0.2-go-modules.drv-0/source/site
> NODE_ENV=production pnpm vite build

failed to load config from /private/tmp/nix-build-coder-2.0.2-go-modules.drv-0/source/site/vite.config.ts
error during build:
Error: The package "@esbuild/darwin-x64" could not be found, and is needed by esbuild.

If you are installing esbuild with npm, make sure that you don't specify the
"--no-optional" or "--omit=optional" flags. The "optionalDependencies" feature
of "package.json" is used by esbuild to install the correct binary executable
for your current platform.
    at generateBinPath (/private/tmp/nix-build-coder-2.0.2-go-modules.drv-0/source/site/node_modules/.pnpm/esbuild@0.18.17/node_modules/esbuild/lib/main.js:1898:15)
    at esbuildCommandAndArgs (/private/tmp/nix-build-coder-2.0.2-go-modules.drv-0/source/site/node_modules/.pnpm/esbuild@0.18.17/node_modules/esbuild/lib/main.js:1968:33)
    at ensureServiceIsRunning (/private/tmp/nix-build-coder-2.0.2-go-modules.drv-0/source/site/node_modules/.pnpm/esbuild@0.18.17/node_modules/esbuild/lib/main.js:2132:25)
    at build (/private/tmp/nix-build-coder-2.0.2-go-modules.drv-0/source/site/node_modules/.pnpm/esbuild@0.18.17/node_modules/esbuild/lib/main.js:2024:26)
    at bundleConfigFile (file:///private/tmp/nix-build-coder-2.0.2-go-modules.drv-0/source/site/node_modules/.pnpm/vite@4.4.2_@types+node@18.17.0/node_modules/vite/dist/node/chunks/dep-8609dc5d.js>
    at loadConfigFromFile (file:///private/tmp/nix-build-coder-2.0.2-go-modules.drv-0/source/site/node_modules/.pnpm/vite@4.4.2_@types+node@18.17.0/node_modules/vite/dist/node/chunks/dep-8609dc5d.>
    at resolveConfig (file:///private/tmp/nix-build-coder-2.0.2-go-modules.drv-0/source/site/node_modules/.pnpm/vite@4.4.2_@types+node@18.17.0/node_modules/vite/dist/node/chunks/dep-8609dc5d.js:65>
    at build (file:///private/tmp/nix-build-coder-2.0.2-go-modules.drv-0/source/site/node_modules/.pnpm/vite@4.4.2_@types+node@18.17.0/node_modules/vite/dist/node/chunks/dep-8609dc5d.js:47740:26)
    at CAC.<anonymous> (file:///private/tmp/nix-build-coder-2.0.2-go-modules.drv-0/source/site/node_modules/.pnpm/vite@4.4.2_@types+node@18.17.0/node_modules/vite/dist/node/cli.js:822:15)
 ELIFECYCLE  Command failed with exit code 1.
/nix/store/w2gc3yd1jgns9mi2f7fscvn9gs83b0ps-stdenv-darwin/setup: line 153: pop_var_context: head of shell_variables not a function context
```

**Comment by chrisportela** [comment]
@carloaromando:

> @chrisportela I tried to build this `coder` version on my `x86_64-darwin` machine but it fails with this log:

It seems like the change purposed above is needed to re-export the ESBUILD path. During my testing I don't think I needed it, but this PR has been open for a while. I am going to find some time to look at this PR more closely again.

**Comment by drupol** [comment]
Hello, I opened a new PR at https://github.com/NixOS/nixpkgs/pull/301512 based on https://github.com/NixOS/nixpkgs/pull/253235#issuecomment-1781338053. Could you please have a look?

**Pull request closed without merging**